### ISSUE TITLE
cucumber html meta data

### ISSUE THREAD

**Issue by a1dutch** [cucumber/cucumber-jvm#2192]
Hi,

How is it possible to pass the product name and product version meta data so its included in the new html report?

I'm using cucumber jvm 6.9.1 and the html plugin.

Cheers

Andy

**Comment by aslakhellesoy** [comment]
These fields are provided by the Cucumber implementation. They are not meant to be set by users. More details here: https://github.com/cucumber/cucumber/blob/master/messages/messages.md#meta

**Comment by a1dutch** [comment]
Hi @aslakhellesoy,

How do we get these populated in the end report?

This is all we see, we would like the product name and version displayed (plus some other custom ones if possible)

![image](https://user-images.githubusercontent.com/1457117/102243155-c13de980-3ef2-11eb-90c6-c549bc2a82bb.png)

**Comment by aslakhellesoy** [comment]
> How do we get these populated in the end report?

Sorry, I don’t understand the question. What do you mean by “these”? Are you talking about the name of the product you are testing/developing with Cucumber?

Do you want to add additional information to the report summary? Can you give me a concrete example (some real values) of what information you would like to add? Could you also explain why you would like to add it?

**Comment by a1dutch** [comment]
In the cucumber messsages docs you posted this has lots of additional information (CI, Git, Product, Version, ...)

![image](https://user-images.githubusercontent.com/1457117/102345493-627d7c00-3f95-11eb-8d31-dd3733a7c60c.png)

Inside cucumber react which is used to generate the html report it should show this information if present: https://github.com/cucumber/cucumber/blob/master/react/javascript/src/components/app/ExecutionSummary.tsx, https://github.com/cucumber/cucumber/blob/master/react/javascript/src/components/app/CICommitLink.tsx

when running from with cucumber-junit via maven none of this information is present or displayed in the html report.

i would like the following displayed at the top of the html report

- Product Name
- Product Version
- Git Commit
- Git Branch
- Git url
- CI url

obviously this information needs to be passed to cucumber when its ran (via command line options or environment variables) for it to be added to the report, but there seems no way at present to do this.

**Comment by aslakhellesoy** [comment]
In the [meta](https://github.com/cucumber/cucumber/blob/master/messages/messages.md#io.cucumber.messages.Meta) message there are 4 fields of type [Meta.Product](https://github.com/cucumber/cucumber/blob/master/messages/messages.md#io.cucumber.messages.Meta.Product):

* implementation
* runtime
* os
* cpu

The `Meta.Product` type is a struct with a `name` and `version`. You can see an example of what this message looks like [here](https://github.com/cucumber/cucumber/blob/master/compatibility-kit/javascript/features/minimal/minimal.ndjson)

Here it is, prettified:

```json
{
  "meta": {
    "protocolVersion": "13.0.1",
    "implementation": {
      "name": "fake-cucumber",
      "version": "8.0.0"
    },
    "runtime": {
      "name": "node.js",
      "version": "12.16.2"
    },
    "os": {
      "name": "linux",
      "version": "4.19.76-linuxkit"
    },
    "cpu": {
      "name": "x64"
    },
    "ci": {
      "name": "GitHub Actions",
      "url": "https://github.com/cucumber-ltd/shouty.rb/actions/runs/154666429",
      "git": {
        "remote": "https://github.com/cucumber-ltd/shouty.rb.git",
        "revision": "99684bcacf01d95875834d87903dcb072306c9ad",
        "branch": "master"
      }
    }
  }
}
```

As you can see, there is no `product` *field* - it is just a type with a `name` and an optional `version` to describe:

* the Cucumber `implementation` (cucumber-ruby, cucumber-js, cucumber-jvm etc)
* the underlying `runtime` (JVM, Node.js, Ruby etc)
* the Operating sustem `os` (Windows, Linux, MacOS etc)
* the CPU architecture `cpu` (x86, amd64 etc)

The `ci` field is a slightly more complex struct of type [Meta.CI](https://github.com/cucumber/cucumber/blob/master/messages/messages.md#metaci) with information about the CI server and the Git remote/revision/branch.

The construction of this entire `meta` message is handled by the [create-meta](https://github.com/cucumber/cucumber/tree/master/create-meta) library, which is used by Cucumber-JVM and other Cucumber implementations.

The `ci` field is populated from [well-known environment variables](https://github.com/cucumber/cucumber/blob/master/create-meta/ciDict.json) defined by the most common CI servers. That's how the Git information is detected as well, as we don't want to shell out to `git` or bundle a native git client.

If no CI environment is detected, the `ci` field will be null, but all the other fields of `meta` should still be defined.

In other words, you have to run Cucumber on one of the [supported CI servers](https://github.com/cucumber/cucumber/tree/master/create-meta#ci) in order for the `ci` field to be populated. 

You could define those environment variables in your terminal manually, but that's error-prone and cumbersome.

In order to simplify generation and publishing of reports we're also offering the [Cucumber Reports](https://reports.cucumber.io/) service, which we recommend using with your favourite CI server.

Does this answer your questions?

**Comment by aslakhellesoy** [comment]
The report screenshot you attached displays all the information you could expect from a non-CI execution. If you run on a CI server you should also see the CI information.

**Comment by a1dutch** [comment]
Thanks that is a lot of information.

We are using jenkins which does not seem supported, you can tell if its running on jenkins by the `JENKINS_URL` environment variable. There are no environment variables for the git information though so reading from a generic environment variable would allow these to be passed in before execution of cucumber.

Being able to display custom fields (product name and version under test, environment under test, profiles, etc...) would make these reports more detailed.  We use the https://github.com/jenkinsci/cucumber-reports-plugin that allows you to do this but we like the single html file that cucumber produces.

The cucumber report service looks nice, but not something we would be able to use due to sensitive data in the reports.
